src/kpro_batch_v01.erl

```diff
@@ -0,0 +1,247 @@
+%%%   You may obtain a copy of the License at
+%%%
+%%%       http://www.apache.org/licenses/LICENSE-2.0
+%%%
+%%%   Unless required by applicable law or agreed to in writing, software
+%%%   distributed under the License is distributed on an "AS IS" BASIS,
+%%%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%%%   See the License for the specific language governing permissions and
+%%%   limitations under the License.
+%%%
+
+% Structures of kafka message v0 and v1
+%
+% v0
+% Message => Offset Size Crc MagicByte Attributes Key Value
+%   Offset => int64
+%   Size => int32
+%   Crc => int32
+%   MagicByte => int8
+%   Attributes => int8
+%   Key => bytes
+%   Value => bytes
+%
+% v1 (supported since 0.10.0)
+% Message => Offset Size Crc MagicByte Attributes Key Value
+%   Offset => int64
+%   Size => int32
+%   Crc => int32
+%   MagicByte => int8
+%   Attributes => int8
+%   Timestamp => int64
+%   Key => bytes
+%   Value => bytes
+
+-module(kpro_batch_v01).
+
+-export([ decode/2
+        , encode/2
+        ]).
+
+-type msg_in() :: kpro:kv() | kpro:tkv().
+-type msg_ts() :: kpro:msg_ts().
+-type key() :: kpro:key().
+-type value() :: kpro:value().
+-type kv_list() :: kpro:kv_list().
+-type compress_option() :: kpro:compress_option().
+-type offset() :: kpro:offset().
+-type message() :: kpro:message().
+
+-include("kpro_private.hrl").
+
+-define(NO_TIMESTAMP, -1).
+
+%%%_* APIs =====================================================================
+
+-spec encode(kv_list(), compress_option()) -> iodata().
+encode(KvList, Compression) ->
+  {Encoded, WrapperTs} = do_encode_messages(KvList),
+  case Compression =:= ?no_compression of
+    true -> Encoded;
+    false -> compress(Compression, Encoded, WrapperTs)
+  end.
+
+%% @doc Decode one message or a compressed batch.
+%% NOTE: Messages are returned in reversed order, so it's cheaper for caller
+%% to concatenate (++) a short header to a long tail.
+%% @end
+%% Comment is copied from:
+%% core/src/main/scala/kafka/message/Message.scala
+%%
+%% The format of an N byte message is the following:
+%% 1. 4 byte CRC32 of the message
+%% 2. 1 byte "magic" identifier to allow format changes, value is 0 or 1
+%% 3. 1 byte "attributes" identifier to allow annotations on the message
+%%           independent of the version
+%%    bit 0 ~ 2 : Compression codec.
+%%      0 : no compression
+%%      1 : gzip
+%%      2 : snappy
+%%      3 : lz4
+%%    bit 3 : Timestamp type
+%%      0 : create time
+%%      1 : log append time
+%%    bit 4 ~ 7 : reserved
+%% 4. (Optional) 8 byte timestamp only if "magic" identifier is greater than 0
+%% 5. 4 byte key length, containing length K
+%% 6. K byte key
+%% 7. 4 byte payload length, containing length V
+%% 8. V byte payload
+-spec decode(offset(), binary()) -> [message()].
+decode(Offset, <<CRC:32/unsigned-integer, Body/binary>>) ->
+  CRC = erlang:crc32(Body), %% assert
```

Comment:
In a previous module you called crc32cer:nif/1 to "assert" the CRC, here you call erlang:crc32/1.  Is there a reason for this difference?

Reply:
kpro_batch_v01 is to handle 'legacy' (magic v0 and v1) batches where classic CRC is used.
in magic v2, it's CRC-32C